File: ga/gpt3_5_result_2024_6/result/multi_thread_basic_gpt/0_multi_thread_basic.c
use std::sync::{Arc, Mutex};
use std::thread;

// Global shared variable representing account balance
static mut ACCOUNT_BALANCE: i32 = 1000000;

fn withdraw() {
    let amount = 500000;
    for _ in 0..amount {
        unsafe {
            ACCOUNT_BALANCE -= 1;
        }
    }
}

fn main() {
    let balance = Arc::new(Mutex::new(ACCOUNT_BALANCE));

    let balance_clone1 = Arc::clone(&balance);
    let handle1 = thread::spawn(move || {
        withdraw();
    });

    let balance_clone2 = Arc::clone(&balance);
    let handle2 = thread::spawn(move || {
        withdraw();
    });

    handle1.join().unwrap();
    handle2.join().unwrap();

    let final_balance = balance.lock().unwrap();
    println!("Final account balance: {}", final_balance);

    // Mutex automatically dropped here
}

File: ga/4omini_result_2024_7/result/multi_thread_basic_mini/0_multi_thread_basic.rs
use std::sync::{Arc, Mutex};
use std::thread;

const INITIAL_BALANCE: i32 = 1000000;
const WITHDRAW_AMOUNT: i32 = 500000;

fn withdraw(account_balance: Arc<Mutex<i32>>, amount: i32) {
    for _ in 0..amount {
        let mut balance = account_balance.lock().unwrap();
        *balance -= 1;  
    }
}

fn main() {
    // Shared account balance wrapped in Arc and Mutex
    let account_balance = Arc::new(Mutex::new(INITIAL_BALANCE));

    // Create two threads for withdrawing
    let account_clone1 = Arc::clone(&account_balance);
    let account_clone2 = Arc::clone(&account_balance);

    let t1 = thread::spawn(move || withdraw(account_clone1, WITHDRAW_AMOUNT));
    let t2 = thread::spawn(move || withdraw(account_clone2, WITHDRAW_AMOUNT));

    // Wait for both threads to finish
    t1.join().unwrap();
    t2.join().unwrap();

    // Print the final account balance
    let final_balance = account_balance.lock().unwrap();
    println!("Final account balance: {}", *final_balance);
}